### DISCUSSION TITLE
How to utilize ipxe with computers without wired network connector?

### DISCUSSION THREAD

**Discussion by Hasky23** [ipxe/ipxe#848] [Q&A] [answered]
Hello,

Nowadays some laptops do not have a connector for network cable and only available option is to use USB NIC interface.

Ipxe for UEFI can be copied on USB drive with UEFI partition, but what is a procedure to access USB NIC interface?

I would guess that somehow ipxe will require to find NIC chip connected to USB port, a quick search in documentation did not gave desired result.

Thanks for any pointers

**Comment by stappersg** [comment]
> Thanks for any pointers

* https://github.com/ipxe/ipxe/discussions?discussions_q=USB
* https://github.com/ipxe/ipxe/issues?q=is%3Aissue+is%3Aopen+USB
* https://github.com/ipxe/ipxe/pulls?q=is%3Apr+USB

plus hardware that has/gets proper software.

**Comment by NiKiZe** [comment]
Many machines have netboot support over USB, iPXE can use the SNP/NII interface on these machines without issues, so you might not even need a iPXE build with USB support.

**Comment by mcb30** [comment, marked as answer]
The default build of iPXE unfortunately does not include native drivers for USB NICs.  This is because of several known bugs in vendor UEFI USB implementations, which can result in a broken system if iPXE's native USB support is included.

iPXE currently has the following USB NIC drivers:

* `acm`
* `axge`
* `dm96xx`
* `ecm`
* `iphone`
* `lan78xx`
* `ncm`
* `smsc75xx`
* `smsc95xx`

You can build a version that includes selected drivers using e.g. `make bin-x86_64-efi/ipxe--ncm--ecm--acm--iphone.efi` or `make bin-x86_64-efi/ipxe--ncm--ecm--acm--iphone.usb`.

**Reply by Hasky23** [reply]
Can you clarify on `iphone` option? Is iphone utilized as a NIC somehow?

**Reply by mcb30** [reply]
The `iphone` driver supports using a USB-attached iPhone via tethering.

Android phones are supported by the `acm` or `ecm` drivers, since Android will generally emulate one of these standard USB NICs.

**Comment by Hasky23** [comment]
Hello,

in an experiment was utilized [USB NIC](https://www.aliexpress.com/item/1005004646857043.html) ([Realtek 8153](https://www.realtek.com/en/products/communications-network-ics/item/rtl8153b-vb-cg)), laptop [HP Elitebook Folio 1040 G2](https://support.hp.com/ca-en/document/c04556462) (no access to a laptop without wired interface) and very small USB drive.

For a test initially ipxe was compiled as 
```
make bin-x86_64-efi/ipxe--ncm--ecm--acm--iphone.usb
dd if=bin-x86_64-efi/ipxe--ncm--ecm--acm--iphone.usb of=/dev/{usb_drive}
```
test boot was successful but quite slow. Next compilation was as following
```
make bin-x86_64-efi/ecm.usb
dd if=bin-x86_64-efi/ecm.usb of=/dev/{usb_drive}
```
This time boot process taken significantly less time, DHCP server responded with correct settings and redirected to FOG server.

For a test was taken test image of HDD (11:40) and then restored back (4:20), boot of the laptop after imaging was successful.

Thanks to everyone who responded to my request for the assistance.
